notebooks/byod.py

```diff
@@ -0,0 +1,224 @@
+    """
+
+with callysto.Cell("markdown"):
+    """
+    ## Organize your upload to Terra
+    Using instructions from the document above, you will use the gsutil tool to upload data from your local computer to
+    your Terra workspace. To effectively use the example here, we suggest you upload the data with the suggestions
+    below.
+    """
+
+with callysto.Cell("markdown"):
+    """
+    ### Find the path to this workspace bucket
+    """
+
+with callysto.Cell("markdown"):
+    """
+    Using the os package, you can print your workspace bucket path.
+    """
+
+with callysto.Cell("python"):
+    bucket = os.environ["WORKSPACE_BUCKET"]
+    print(bucket)
+
+with callysto.Cell("markdown"):
+    """
+    ### Add a prefix to your bucket path to organize your data
+    In this example, we add the prefix 'test-crai-cram'. In the terminal of your computer, you will call something like:
+
+    `gsutil cp /Users/Documents/Example.cram gs://your_bucket_info/test-crai-cram/`
+    """
+
+with callysto.Cell("markdown"):
+    """
+    ## Preview the data in your workspace bucket
+    """
+
+with callysto.Cell("python"):
+    #!gsutil ls {bucket}
+    pass
+
+with callysto.Cell("markdown"):
+    """
+    # Generate a data table that links to the data in your workspace bucket
+    """
+
+################################################ TESTS ################################################ noqa
+delete_table("test_cram_crai_table")
+listing = list()
+for i in range(5):
+    listing.append(f"gs://some-bucket/some-pfx/sample_id_{i}.cram")
+    listing.append(f"gs://some-bucket/some-pfx/sample_id_{i}.crai")
+create_cram_crai_table("test_cram_crai_table", listing)
+keyed_rows = get_keyed_rows("test_cram_crai_table", "sample")
+for i in range(5):
+    assert keyed_rows[f'sample_id_{i}'] == dict(cram=f"gs://some-bucket/some-pfx/sample_id_{i}.cram",
+                                                crai=f"gs://some-bucket/some-pfx/sample_id_{i}.crai")
+
+delete_table("test_metadata_table_a")
+test_metadata_table_a_columns = dict(sample=["sample_id_1", "sample_id_2", "sample_id_3", "sample_id_4"],
+                                     firstname=["a", "b", "c", "d"],
+                                     birthday=["e", "f", "g", "h"])
+upload_columns("test_metadata_table_a", test_metadata_table_a_columns)
+
+delete_table("test_metadata_table_b")
+test_metadata_table_b_columns = dict(sample=["sample_id_1", "sample_id_2", "sample_id_3", "sample_id_5"],
+                                     alpha=["1", "2", "3", "4"],
+                                     beta=["5", "6", "7", "8"])
+upload_columns("test_metadata_table_b", test_metadata_table_b_columns)
+
+delete_table("test_joined_table")
+join_data_tables("test_joined_table", ["test_cram_crai_table", "test_metadata_table_a", "test_metadata_table_b"], "sample")
```

Comment:
If I understand what you're doing here we've definitely duplicated efforts. Let's see if we can merge our notebooks.

Reply:
This is a general method that will work for any Terra data tables with columns of shared values, whether or not they were imported from Gen3.